### PR TITLE
fix(source): introduce state_table_id as state store prefix 

### LINKED ISSUE #4817: Use a separate table id for each source executor
**Describe the bug**
Currently we use `table_id` and `split_id` to distinguish the internal state of each source executor, which is usually fine, but in the case of a source joining itself, it will lead to a key conflict in the state, and the temporary solution is to avoid it by adding `fragment_id`, but this is not a long-term plan, so we need to create a separate internal table id for each source executor

**To Reproduce**
Steps to reproduce the behavior.

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by tabVersion**
Signed-off-by: tabVersion <tabvision@bupt.icu>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

as title

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

fix #4817

### CODE REVIEW

**Review by BugenZhao — COMMENTED** [review]
Generally LGTM. Do we plan to apply state table for source executor state in the future? ❤️ cc @st1page @yezizp2012

File: src/meta/src/stream/stream_graph.rs [hunk 1]
```diff
@@ -568,6 +568,14 @@ impl StreamGraphBuilder {
                         }
                     }
 
+                    NodeBody::Source(node) => update_table(
+                        &mut Table {
+                            id: node.state_table_id,
+                            ..Default::default()
+                        },
+                        "SourceNode",
+                    ),
```
Comment (yezizp2012) [inline src/meta/src/stream/stream_graph.rs:577]:
```suggestion
                    NodeBody::Source(node) => {
                        node.state_table_id += table_id_offset;
                        // fill internal table for source node with None catalog.
                        check_and_fill_internal_table(node.state_table_id, None);
                    },
```

**Review by yezizp2012 — COMMENTED** [review]
Rest LGTM, the panic should cause by here.

**Comment by tabVersion** [comment]
> Do we plan to apply state table for source executor state in the future?

I think no need for that. Source just uses state store as a KV

File: proto/stream_plan.proto [hunk 2]
```diff
@@ -107,9 +107,10 @@ message SourceNode {
     TABLE = 1;
     SOURCE = 2;
   }
-  uint32 table_id = 1;
+  uint32 table_id = 1; // use table_id / source_id to fetch SourceDesc from local source manager
```
Comment (yezizp2012) [inline proto/stream_plan.proto:110]:
```suggestion
  uint32 source_id = 1; // use source_id to fetch SourceDesc from local source manager
```

Comment (yezizp2012) [reply proto/stream_plan.proto:110]: `source_id` seems more appropriate here, i'm okay to keep or rename it.

Comment (Li0k) [inline proto/stream_plan.proto:110]: Is it possible keep using `table_id` to access `state_store`, and rename now `table_id` to other in order to keep the concept with other operators

**Review by yezizp2012 — APPROVED** [review]
LGTM.

**Review by Li0k — APPROVED** [review]
LGTM

File: src/frontend/src/optimizer/plan_node/stream_source.rs [hunk 3]
```diff
@@ -76,6 +76,7 @@ impl ToStreamProst for StreamSource {
                 .map(|c| c.column_id().into())
                 .collect(),
             source_type: self.logical.source_catalog.source_type as i32,
+            state_table_id: 0,
```
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/stream_source.rs:79]:
```suggestion
            state_table_id: TableId::placeholder(),
```

**Review by st1page — APPROVED** [review]
rest LGTM, will the modification of source executor in the next PR?

File: src/frontend/src/optimizer/plan_node/stream_source.rs (hunk 3 above)
Comment (BugenZhao) [reply src/frontend/src/optimizer/plan_node/stream_source.rs:79]: +1

**Comment by BugenZhao** [comment]
> > Do we plan to apply state table for source executor state in the future?
> 
> I think no need for that. Source just uses state store as a KV

Actually, the storage side may decide not to provide a raw K-V interface for streaming usage in the future. I'll take care of this. 🤣

**Comment by tabVersion** [comment]
> > > Do we plan to apply state table for source executor state in the future?
> > 
> > 
> > I think no need for that. Source just uses state store as a KV
> 
> Actually, the storage side may decide not to provide a raw K-V interface for streaming usage in the future. I'll take care of this. 🤣

I think it is a little off the topic. Source executor may need an extra field for split assigned but not consumed (#4821). If it has to be a table, I think we can make it like `(prev_state varchar, new_assigned varchar)`

**Pull request merged**